### DISCUSSION TITLE
int.Parse("1_000") should work

### DISCUSSION THREAD

**Discussion by ed-graham** [dotnet/runtime#79863] [General] [answered]
Since C#7, `int i = 1_000;` is valid syntax; yet `int.Parse("1_000");` will still fail.  This seems inconsistent.

I know I could do it with `NumberFormatInfo.NumberGroupSeparator` but that feels relatively painful.

**Comment by The-Futurist** [comment]
> Since C#7, `int i = 1_000;` is valid syntax; yet `int.Parse("1_000");` will still fail. This seems inconsistent.
> 
> I know I could do it with `NumberFormatInfo.NumberGroupSeparator` but that feels relatively painful.

I suspect that is not an error here. That `1_000` is a notation specific to the language, supported and recognized only by the lexer/parser.

It's an interesting point though, perhaps there is some other supported way to do this...

I misinterpreted then, you wrote "should work" as if it wasn't correct, not working, I see what you mean now!

**Reply by ed-graham** [reply]
Sorry @Korporal , I don't fully understand what you mean when you say it's not an error.  I realise this is intended behaviour at present; but I thought this was a forum for suggesting improvements to the language.

**Reply by ed-graham** [reply]
Ah yes - my title was misleading, I see.  I've tried to change it but I think it's now too late!

**Comment by HaloFour** [comment, marked as answer]
This would be a request for the [runtime](https://github.com/dotnet/runtime/) as it would be an API change.

**Reply by tannergooding** [reply]
It's also likely not something we would take on the runtime side. This is really language specific support and would be akin to supporting `float.Parse("1000.0f")`.

Today, we largely only support locale specific separates (such as `.` or `,`). Supporting other factors like `_` would be a behavioral change that could have undesirable side effects and which would add overhead to the processing of an otherwise fairly perf critical piece of code.

**Reply by ed-graham** [reply]
Thanks @HaloFour and @tannergooding.  I made the suggestion because we display values on a log scale in our app (1000, 10000, 100000 etc.) in a drop-down, and our users asked for the _ separator to make them clearer.  Before the change I just had `return int.Parse(valueAsString);` as my `SelectedIndexChanged` event handler; but now I have a `switch` statement, which is a lot more cumbersome.

**Reply by En3Tho** [reply]
Doesn't valueAsString.Replace("_", "") solve your problem?

**Reply by Clockwork-Muse** [reply]
Or creating a custom locale with that as a separator.
Or just using their actual local separator?  Was there some reason they wanted something unusual?

**Reply by ed-graham** [reply]
> Doesn't valueAsString.Replace("_", "") solve your problem?

Why, yes; yes it does!  I was clearly lacking imagination :)  I was focusing on the fact that the language didn't support it rather than finding an elegant work-around.  This is indeed much neater than a `switch` statement - thanks!

**Reply by ed-graham** [reply]
> Or creating a custom locale with that as a separator. Or just using their actual local separator? Was there some reason they wanted something unusual?

Yes, I could certainly create a custom locale.  I was looking for something a bit quicker and easier; but @En3Tho's answer above will do the trick.

**Reply by ed-graham** [reply]
Thanks to all who replied.  I see now that this is harder than it first appears, and also that there is a simple yet elegant work-around available, so I'll mark this as resolved and forever hold my peace.
